### ISSUE TITLE
Crash when there are assembly loading error with ByRef-like type cannot be used as the type for an instance field in a non-ByRef-like type

### ISSUE THREAD

**Issue by 304NotModified** [AdCodicem/SpecFlowMicrosoftDependencyInjection#9]
I get this error, ~~but only on my CI~~ it on;y works with the Resharper runner and is now crashing in VS runner and the CI

Any idea why?

```
[xUnit.net 00:00:02.21]   Starting:    Namespace1.Application1.AcceptanceTests
-> Loading plugin /LivingDoc.SpecFlowPlugin.dll
-> Loading plugin /TechTalk.SpecFlow.xUnit.SpecFlowPlugin.dll
-> Loading plugin /AdCodicem.SpecFlow.MicrosoftDependencyInjection.SpecFlowPlugin.dll
-> Loading plugin /Namespace1.Application1.AcceptanceTests.dll
-> Using specflow.json
-> #16: warning: The previous ScenarioContext was already disposed.
[xUnit.net 00:00:02.68]       System.AggregateException : One or more errors occurred. (Unable to load one or more of the requested types.
[xUnit.net 00:00:02.68]       A ByRef-like type cannot be used as the type for an instance field in a non-ByRef-like type.) (Object reference not set to an instance of an object.)[xUnit.net 00:00:02.67]     My test [FAIL]

[xUnit.net 00:00:02.68]       ---- System.Reflection.ReflectionTypeLoadException : Unable to load one or more of the requested types.
[xUnit.net 00:00:02.68]       A ByRef-like type cannot be used as the type for an instance field in a non-ByRef-like type.
[xUnit.net 00:00:02.68]       ---- System.NullReferenceException : Object reference not set to an instance of an object.
[xUnit.net 00:00:02.68]       Stack Trace:
[xUnit.net 00:00:02.68]         
[xUnit.net 00:00:02.68]         ----- Inner Stack Trace #1 (System.Reflection.ReflectionTypeLoadException) -----
[xUnit.net 00:00:02.68]            at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
[xUnit.net 00:00:02.68]            at System.Reflection.RuntimeModule.GetTypes()
[xUnit.net 00:00:02.68]            at System.Reflection.Assembly.GetTypes()
[xUnit.net 00:00:02.68]            at AdCodicem.SpecFlow.MicrosoftDependencyInjection.DependencyInjectionPlugin.<>c.<ConfigureBindings>b__1_0(Assembly assembly)
[xUnit.net 00:00:02.68]            at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
[xUnit.net 00:00:02.68]            at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
[xUnit.net 00:00:02.68]            at AdCodicem.SpecFlow.MicrosoftDependencyInjection.DependencyInjectionPlugin.ConfigureBindings(IServiceCollection services)
[xUnit.net 00:00:02.68]            at AdCodicem.SpecFlow.MicrosoftDependencyInjection.DependencyInjectionPlugin.ConfigureServices(IObjectContainer delegateContainer)
[xUnit.net 00:00:02.68]            at AdCodicem.SpecFlow.MicrosoftDependencyInjection.DependencyInjectionPlugin.<Initialize>b__3_1(Object _, CustomizeScenarioDependenciesEventArgs args)
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.Plugins.RuntimePluginEvents.RaiseCustomizeScenarioDependencies(ObjectContainer scenarioContainer)
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.Infrastructure.ContainerBuilder.CreateScenarioContainer(IObjectContainer testThreadContainer, ScenarioInfo scenarioInfo)
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.Infrastructure.ContextManager.InitializeScenarioContext(ScenarioInfo scenarioInfo)
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.OnScenarioInitialize(ScenarioInfo scenarioInfo)
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.TestRunner.OnScenarioInitialize(ScenarioInfo scenarioInfo)
[xUnit.net 00:00:02.68]            at Namespace1.Application1.AcceptanceTests.End2End.Feature1.ScenarioInitialize(ScenarioInfo scenarioInfo)
[xUnit.net 00:00:02.68]         /agent/_work/1/s/Backend/Namespace1.Application1.AcceptanceTests/End2End/My test.feature(4,0): at Namespace1.Application1.AcceptanceTests.End2End.Feature1.MyTest()
[xUnit.net 00:00:02.68]         ----- Inner Stack Trace #2 (System.NullReferenceException) -----
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.OnScenarioEnd()
[xUnit.net 00:00:02.68]            at TechTalk.SpecFlow.TestRunner.OnScenarioEnd()
[xUnit.net 00:00:02.68]            at Namespace1.Application1.AcceptanceTests.End2End.Feature1.TestTearDown()
[xUnit.net 00:00:02.68]            at Namespace1.Application1.AcceptanceTests.End2End.Feature1.System.IDisposable.Dispose()
[xUnit.net 00:00:02.68]         C:\Dev\xunit\xunit\src\xunit.execution\Extensions\ReflectionAbstractionExtensions.cs(79,0): at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)
```

**Comment by 304NotModified** [comment]
Funny, it's because of IronPDF. I cannot remove that dependency

![image](https://user-images.githubusercontent.com/5808377/157767584-fcd3889e-bea3-4830-92d3-4ef71248181d.png)

**Comment by 304NotModified** [comment]
From the screenshot above it looks like a VS or xUnit problem.

But if I remove the package, then the test is green and won't crash

**Comment by AdCodicem** [comment]
Hello,
From what I see in your screenshot, I'm not sure that the DI plugin caused the issue.
Can you provide me a link to a repo presenting the issue, so I can investigate ? Thanks !

**Comment by 304NotModified** [comment]
The issue is here:

```
[xUnit.net 00:00:02.68]            at System.Reflection.RuntimeModule.GetTypes()
[xUnit.net 00:00:02.68]            at System.Reflection.Assembly.GetTypes()
[xUnit.net 00:00:02.68]            at AdCodicem.SpecFlow.MicrosoftDependencyInjection.DependencyInjectionPlugin.<>c.<ConfigureBindings>b__1_0(Assembly assembly)
```

The issue is that  `System.Reflection.Assembly.GetTypes()` failed, and this is triggered by MicrosoftDependencyInjection.DependencyInjectionPlugin.

I don't think this is easy to reproduce in a small application.

**Comment by 304NotModified** [comment]
So this is 

https://github.com/AdCodicem/SpecFlowMicrosoftDependencyInjection/blob/0c18d04280683e8231ed367524e3a33f1257539f/AdCodicem.SpecFlow.MicrosoftDependencyInjection/DependencyInjectionPlugin.cs#L23-L28

Is there a reason we walk over all the current assemblies in the appdomain? I think this should only be the executing assembly?

So maybe 

Assembly.GetEntryAssembly, Assembly.GetCallingAssembly or Assembly.GetExecutingAssembly. See 
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly?view=net-6.0#methods

**Comment by AdCodicem** [comment]
Thanks for providing more info.
I think this is related to this issue : #6
That's why I checked all the assemblies.
I think I can solve this issue by filtering the types.
I'll try to reproduce this issue.

**Comment by AdCodicem** [comment]
Hi, since I don't have access to your code, can you test this for me ?
- Download the DI plugin project and register it directly as a dependency instead of using the nuget.
- In the code snippet you identified earlier, replace `assembly.GetTypes()` with `assembly.GetLoadableTypes()` 
- Run your tests again and tell me if the issue is fixed 
Thanks !

**Comment by 304NotModified** [comment]
Yes I could test, but maybe a pre-release package would be easier / better?

**Comment by AdCodicem** [comment]
Hello, as you suggested, I published a pre-release package :
https://www.nuget.org/packages/AdCodicem.SpecFlow.MicrosoftDependencyInjection/1.1.6-cmarty-fix-by-ref-type-registration.1

**Comment by 304NotModified** [comment]
thanks!

**Comment by 304NotModified** [comment]
Unfortunately this won't fix it. 

GetLoaderTypes just rethrows this error:

![image](https://user-images.githubusercontent.com/5808377/157856275-1f49f77c-fb05-4d3d-b472-9ad7cfbada6f.png)

![image](https://user-images.githubusercontent.com/5808377/157856309-a3d73659-abb5-48bd-bb93-1e7fd36703e4.png)

But if we check the code of the VSTest runner (decompiled sources), it just ignores these errors, see the `continue` in the `catch`
The same error happens on `assembly.DefinedTypes`.

```c#
            foreach (var assemblyName in depsAssms.Concat(dllsInSources))
            {
                try
                {
                    var assembly = Assembly.Load(assemblyName);
                    foreach (var type in assembly.DefinedTypes)
                    {
#pragma warning disable CS0618
                        if (type == null || type.IsAbstract || type == typeof(DefaultRunnerReporter).GetTypeInfo() || type == typeof(DefaultRunnerReporterWithTypes).GetTypeInfo() || type.ImplementedInterfaces.All(i => i != typeof(IRunnerReporter)))
                            continue;
#pragma warning restore CS0618

                        var ctor = type.DeclaredConstructors.FirstOrDefault(c => c.GetParameters().Length == 0);
                        if (ctor == null)
                        {
                            ConsoleHelper.SetForegroundColor(ConsoleColor.Yellow);
                            Console.WriteLine($"Type {type.FullName} in assembly {assembly} appears to be a runner reporter, but does not have an empty constructor.");
                            ConsoleHelper.ResetColor();
                            continue;
                        }

                        result.Add((IRunnerReporter)ctor.Invoke(new object[0]));
                    }
                }
                catch
                {
                    continue;
                }
```

So proposal: also ignore (and maybe log them to Debug), these loaders errors (of course move to the next assemby)

**Comment by 304NotModified** [comment]
related in NLog ;)

https://github.com/NLog/NLog/blob/0b952cde2a2bda471663695788b5d3a626fb8b41/src/NLog/Internal/Reflection/ReflectionHelpers.cs#L48-L85


```c#
        /// <summary>
        /// Gets all usable exported types from the given assembly.
        /// </summary>
        /// <param name="assembly">Assembly to scan.</param>
        /// <returns>Usable types from the given assembly.</returns>
        /// <remarks>Types which cannot be loaded are skipped.</remarks>
        public static Type[] SafeGetTypes(this Assembly assembly)
        {
            try
            {
                return assembly.GetTypes();
            }
            catch (ReflectionTypeLoadException typeLoadException)
            {
                foreach (var ex in typeLoadException.LoaderExceptions)
                {
                    InternalLogger.Warn(ex, "Type load exception.");
                }

                var loadedTypes = new List<Type>();
                foreach (var t in typeLoadException.Types)
                {
                    if (t != null)
                    {
                        loadedTypes.Add(t);
                    }
                }

                return loadedTypes.ToArray();
            }
            catch (Exception ex)
            {
                InternalLogger.Warn(ex, "Type load exception.");
                return ArrayHelper.Empty<Type>();
            }
        }
```

**Comment by 304NotModified** [comment]
See also 
https://stackoverflow.com/questions/7889228/how-to-prevent-reflectiontypeloadexception-when-calling-assembly-gettypes

TL;DR catch it;

I like this simple approach: https://stackoverflow.com/a/65023201/201303

**Comment by AdCodicem** [comment]
I don't think it'll solve the issue, because this is exactly what it's done in the `GetLoadableTypes` extension method

https://github.com/AdCodicem/SpecFlowMicrosoftDependencyInjection/blob/0c18d04280683e8231ed367524e3a33f1257539f/AdCodicem.SpecFlow.MicrosoftDependencyInjection/Internal/AssemblyExtensions.cs#L17-L24

I'm still investigating

**Comment by AdCodicem** [comment]
So I tried to exclude the "By-Ref" types from being loaded, here is another pre-release package: 
https://www.nuget.org/packages/AdCodicem.SpecFlow.MicrosoftDependencyInjection/1.1.6-cmarty-fix-by-ref-type-registration.2

Can you test it and tell me if it solved your error ?

**Comment by 304NotModified** [comment]
> I don't think it'll solve the issue, because this is exactly what it's done in the `GetLoadableTypes` extension method
> 
> https://github.com/AdCodicem/SpecFlowMicrosoftDependencyInjection/blob/0c18d04280683e8231ed367524e3a33f1257539f/AdCodicem.SpecFlow.MicrosoftDependencyInjection/Internal/AssemblyExtensions.cs#L17-L24
> 
> I'm still investigating

ow yes indeed. The decompiled source was a bit hard to read haha

**Comment by 304NotModified** [comment]
I'm busy testing. First signs are good :)

I'm curious what the changes are

**Comment by 304NotModified** [comment]
PS if you merge #11 and #12, it's easier to debug :)

**Comment by 304NotModified** [comment]
> So I tried to exclude the "By-Ref" types from being loaded, here is another pre-release package: https://www.nuget.org/packages/AdCodicem.SpecFlow.MicrosoftDependencyInjection/1.1.6-cmarty-fix-by-ref-type-registration.2
> 
> Can you test it and tell me if it solved your error ?

Confirm, this works!

**Comment by AdCodicem** [comment]
Fixed by PR #13

**Comment by 304NotModified** [comment]
Thanks!
